tokio/src/task/task_local.rs

```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
     /// });
     /// # }
     /// ```
+    ///
+    /// [`with`]: fn@Self::with
+    /// [`try_with`]: fn@Self::try_with
+    #[track_caller]
     pub fn sync_scope<F, R>(&'static self, value: T, f: F) -> R
     where
         F: FnOnce() -> R,
     {
-        let scope = TaskLocalFuture {
-            local: self,
-            slot: Some(value),
-            future: (),
-            _pinned: PhantomPinned,
-        };
-        crate::pin!(scope);
-        scope.with_task(|_| f())
+        let mut value = Some(value);
+        match self.scope_inner(&mut value, f) {
+            Ok(res) => res,
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("sync_scope called while Task Local Storage is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
```

Comment:
very tiny nit: in the docs, we write "task-local storage", but in the panics, we write "Task Local Storage". i have a slight preference for keeping this consistent:
```suggestion
                panic!("sync_scope called while task-local storage is borrowed")
            }
            Err(ScopeInnerErr::AccessError) => {
                panic!("cannot access a task-local storage value during or after destruction")
```